File: app/src/main/java/com/amazon/tv/leanbacklauncher/apps/LaunchPoint.java
package com.amazon.tv.leanbacklauncher.apps;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.pm.ActivityInfo;
import android.content.pm.PackageManager;
import android.content.pm.PackageManager.NameNotFoundException;
import android.content.pm.ResolveInfo;
import android.content.res.Resources;
import android.content.res.Resources.Theme;
import android.content.res.TypedArray;
import android.graphics.Bitmap;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.support.v4.content.ContextCompat;
import android.support.v4.content.res.ResourcesCompat;
import android.support.v7.graphics.Palette;
import android.text.TextUtils;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.bumptech.glide.request.RequestOptions;
import com.bumptech.glide.request.target.SimpleTarget;
import com.bumptech.glide.request.transition.Transition;
import com.amazon.tv.leanbacklauncher.R;
import com.amazon.tv.leanbacklauncher.util.Util;
import com.amazon.tv.firetv.leanbacklauncher.util.AppCategorizer;
import com.amazon.tv.firetv.leanbacklauncher.apps.AppCategory;
import com.amazon.tv.firetv.leanbacklauncher.util.BannerUtil;
import com.amazon.tv.firetv.leanbacklauncher.util.SettingsUtil;

import java.util.Map;

public class LaunchPoint {
    private String mAppTitle;
    private Drawable mBannerDrawable;
    private String mComponentName;
    private String mContentDescription;
    private boolean mHasBanner;
    private Drawable mIconDrawable;
    private int mInstallProgressPercent;
    private int mInstallStateStringResourceId;
    private boolean mIsInitialInstall;
    private int mLaunchColor;
    private Intent mLaunchIntent;
    private InstallingLaunchPointListener mListener;
    private long mPackageInstallTime;
    private String mPackageName;
    private int mPriority;
    private int mSettingsType = SettingsUtil.SettingsType.UNKNOWN.getCode();
    private boolean mTranslucentTheme;

    private boolean mIsSettingsItem;
    private AppCategory mAppCategory;
    private boolean hasGameFlag;

    public AppCategory getAppCategory() {
        return mAppCategory;
    }

    public LaunchPoint() {

    }

    LaunchPoint(Context context, String appTitle, String packageName) {
        clear(context);
        this.mAppTitle = appTitle;
        this.mPackageName = packageName;
    }

    public LaunchPoint(Context context, String appTitle, String iconUrl, String pkgName, Intent launchIntent, InstallingLaunchPointListener listener) {
        this.mAppTitle = appTitle;
        Resources resources = context.getResources();
        this.mLaunchColor = ResourcesCompat.getColor(resources, R.color.app_launch_ripple_default_color, null);
        this.mPackageName = pkgName;
        this.mPackageInstallTime = Util.getInstallTimeForPackage(context, this.mPackageName);
        if (launchIntent != null) {
            this.mLaunchIntent = launchIntent.addFlags(270532608);
            if (this.mLaunchIntent.getComponent() != null) {
                this.mComponentName = this.mLaunchIntent.getComponent().flattenToString();
            }
        }
        this.mListener = listener;
        this.mIsInitialInstall = true;
    }

    public LaunchPoint(Context context, String appTitle, String iconUrl, Intent launchIntent, int launchColor, InstallingLaunchPointListener listener) {
        clear(context);
        this.mAppTitle = appTitle;

        this.mLaunchColor = launchColor;
        if (launchIntent != null) {
            this.mLaunchIntent = launchIntent.addFlags(270532608);
            if (this.mLaunchIntent.getComponent() != null) {
                this.mComponentName = this.mLaunchIntent.getComponent().flattenToString();
                this.mPackageName = this.mLaunchIntent.getComponent().getPackageName();
            }
        }

        Resources resources = context.getResources();
        this.mLaunchColor = ResourcesCompat.getColor(resources, R.color.app_launch_ripple_default_color, null);
        this.mPackageInstallTime = Util.getInstallTimeForPackage(context, this.mPackageName);

        this.mListener = listener;
        this.mIsInitialInstall = true;

        if (!TextUtils.isEmpty(iconUrl)) {
            int maxIconSize = resources.getDimensionPixelSize(R.dimen.banner_icon_size);
            Glide.with(context).asDrawable().apply(RequestOptions.skipMemoryCacheOf(true)).apply(RequestOptions.diskCacheStrategyOf(DiskCacheStrategy.NONE)).load(iconUrl).into(new SimpleTarget<Drawable>(maxIconSize, maxIconSize) {
                public void onResourceReady(Drawable drawable, Transition<? super Drawable> transition) {
                    InstallingLaunchPointListener launchPointListener = LaunchPoint.this.mListener;
                    if (drawable != null && launchPointListener != null) {
                        LaunchPoint.this.mIconDrawable = drawable;
                        launchPointListener.onInstallingLaunchPointChanged(LaunchPoint.this);
                    }
                }
            });
        }
    }

    // todo
    public LaunchPoint(Context context, String appTitle, Drawable iconDrawable, Intent launchIntent, int launchColor) {
        clear(context);
        this.mAppTitle = appTitle;
        this.mIconDrawable = iconDrawable;
        this.mLaunchColor = launchColor;
        if (launchIntent != null) {
            this.mLaunchIntent = launchIntent.addFlags(270532608);
            if (this.mLaunchIntent.getComponent() != null) {
                this.mComponentName = this.mLaunchIntent.getComponent().flattenToString();
                this.mPackageName = this.mLaunchIntent.getComponent().getPackageName();
                this.mPackageInstallTime = Util.getInstallTimeForPackage(context, this.mPackageName);
            }
        }
    }


    public LaunchPoint(Context ctx, PackageManager pm, ResolveInfo info) {
        set(ctx, pm, info);
        this.mIsSettingsItem = true;
    }

    public LaunchPoint(Context ctx, PackageManager pm, ResolveInfo info, boolean useBanner, int settingsType) {
        set(ctx, pm, info, useBanner);
        this.mSettingsType = settingsType;
        this.mIsSettingsItem = true;
    }

    public LaunchPoint set(Context ctx, PackageManager pm, ResolveInfo info) {
        return set(ctx, pm, info, true);
    }

    public LaunchPoint set(Context ctx, PackageManager pm, ResolveInfo info, boolean useBanner) {
        clear(ctx);
        this.mAppTitle = info.loadLabel(pm).toString();
        this.mLaunchIntent = getLaunchIntent(info);
        if (this.mLaunchIntent.getComponent() != null) {
            this.mComponentName = this.mLaunchIntent.getComponent().flattenToString();
            this.mPackageName = this.mLaunchIntent.getComponent().getPackageName();
        }

        Resources res = ctx.getResources();
        int maxWidth = res.getDimensionPixelOffset(R.dimen.max_banner_image_width);
        int maxHeight = res.getDimensionPixelOffset(R.dimen.max_banner_image_height);

        ActivityInfo actInfo = info.activityInfo;
        if (actInfo != null) {
            mAppCategory = AppCategorizer.getAppCategory(this.mPackageName, actInfo);

            if (useBanner) {
                this.mBannerDrawable = actInfo.loadBanner(pm);
                if (this.mBannerDrawable instanceof BitmapDrawable) {
                    this.mBannerDrawable = new BitmapDrawable(res, Util.getSizeCappedBitmap(((BitmapDrawable) this.mBannerDrawable).getBitmap(), maxWidth, maxHeight));
                }
            }

            Map<String, Integer> overrides = BannerUtil.BANNER_OVERRIDES;

            for (String str : overrides.keySet()) {
                if (this.mPackageName.toLowerCase().contains(str)) {
                    this.mBannerDrawable = ContextCompat.getDrawable(ctx, overrides.get(str));
                    if (this.mBannerDrawable instanceof BitmapDrawable) {
                        this.mBannerDrawable = new BitmapDrawable(res, Util.getSizeCappedBitmap(((BitmapDrawable) this.mBannerDrawable).getBitmap(), maxWidth, maxHeight));
                    }
                    break;
                }
            }

            if (this.mBannerDrawable != null) {
                this.mHasBanner = true;
            } else {
                if (useBanner) {
                    this.mBannerDrawable = actInfo.loadLogo(pm);
                    if (this.mBannerDrawable instanceof BitmapDrawable) {
                        this.mBannerDrawable = new BitmapDrawable(res, Util.getSizeCappedBitmap(((BitmapDrawable) this.mBannerDrawable).getBitmap(), maxWidth, maxHeight));
                    }
                }
                if (this.mBannerDrawable != null) {
                    this.mHasBanner = true;
                } else {
                    this.mHasBanner = false;
                    this.mIconDrawable = info.loadIcon(pm);
                }
            }
        }
        this.mPriority = info.priority;
        this.mTranslucentTheme = isTranslucentTheme(ctx, info);
        this.mLaunchColor = getColor(ctx, info);
        this.mPackageInstallTime = Util.getInstallTimeForPackage(ctx, this.mPackageName);
        return this;
    }

    public void addLaunchIntentFlags(int flags) {
        if (this.mLaunchIntent != null) {
            this.mLaunchIntent.addFlags(flags);
        }
    }

    private void clear(Context context) {
        cancelPendingOperations(context);
        this.mInstallProgressPercent = -1;
        this.mInstallStateStringResourceId = 0;
        this.mComponentName = null;
        this.mPackageName = null;
        this.mBannerDrawable = null;
        this.mAppTitle = null;
        this.mContentDescription = null;
        this.mIconDrawable = null;
        this.mLaunchColor = 0;
        this.mLaunchIntent = null;
        this.mHasBanner = false;
        this.mPriority = -1;
        this.mSettingsType = SettingsUtil.SettingsType.UNKNOWN.getCode();
        this.mIsInitialInstall = false;
        this.mListener = null;
        this.mPackageInstallTime = -1;
    }

    public LaunchPoint setInstallationState(LaunchPoint launchPoint) {
        this.mInstallProgressPercent = launchPoint.getInstallProgressPercent();
        this.mInstallStateStringResourceId = launchPoint.getInstallStateStringResId();
        return this;
    }

    private static Intent getLaunchIntent(ResolveInfo info) {
        ComponentName componentName = new ComponentName(info.activityInfo.applicationInfo.packageName, info.activityInfo.name);
        Intent intent = new Intent("android.intent.action.MAIN");
        intent.addCategory("android.intent.category.LAUNCHER");
        intent.setComponent(componentName);
        intent.addFlags(270532608);
        return intent;
    }

    private static int getColor(Context myContext, ResolveInfo info) {
        try {
            Theme theme = myContext.createPackageContext(info.activityInfo.applicationInfo.packageName, 0).getTheme();
            theme.applyStyle(info.activityInfo.getThemeResource(), true);
            TypedArray a = theme.obtainStyledAttributes(new int[]{android.R.attr.colorPrimary});
            int color = a.getColor(0, myContext.getResources().getColor(R.color.banner_background));
            a.recycle();
            return color;
        } catch (NameNotFoundException e) {
            e.printStackTrace();
            return myContext.getResources().getColor(R.color.banner_background);
        }
    }

    private static int getColor(Context context, Bitmap bitmap) {
        Palette p = Palette.from(bitmap).generate();
        Palette.Swatch swatch = p.getDarkMutedSwatch();

        if (swatch != null)
            return swatch.getRgb();
        else if ((swatch = p.getLightMutedSwatch()) != null)
            return swatch.getRgb();
        else if ((swatch = p.getVibrantSwatch()) != null)
            return swatch.getRgb();
        else if ((swatch = p.getMutedSwatch()) != null)
            return swatch.getRgb();
        else if ((swatch = p.getDarkVibrantSwatch()) != null)
            return swatch.getRgb();
        else if ((swatch = p.getDarkMutedSwatch()) != null)
            return swatch.getRgb();

        return ContextCompat.getColor(context, R.color.banner_background); // for API 22
    }

    public static boolean isTranslucentTheme(Context myContext, ResolveInfo info) {
        try {
            Theme theme = myContext.createPackageContext(info.activityInfo.applicationInfo.packageName, 0).getTheme();
            theme.applyStyle(info.activityInfo.getThemeResource(), true);
            TypedArray a = theme.obtainStyledAttributes(new int[]{android.R.attr.windowIsTranslucent});
            boolean windowIsTranslucent = a.getBoolean(0, false);
            a.recycle();
            return windowIsTranslucent;
        } catch (NameNotFoundException e) {
            e.printStackTrace();
            return false;
        }
    }

    public String getComponentName() {
        return this.mComponentName;
    }

    public String getTitle() {
        return this.mAppTitle;
    }

    public void setContentDescription(String desc) {
        this.mContentDescription = desc;
    }

    public String getContentDescription() {
        return this.mContentDescription;
    }

    public boolean hasBanner() {
        return this.mHasBanner;
    }

    public Intent getLaunchIntent() {
        return this.mLaunchIntent;
    }

    public Drawable getBannerDrawable() {
        return this.mBannerDrawable;
    }

    public Drawable getIconDrawable() {
        return this.mIconDrawable;
    }

    public boolean isGame() {
        return this.mAppCategory == AppCategory.GAME;
    }

    public void setIconDrawable(Drawable drawable) {
        this.mIconDrawable = drawable;
    }

    public void setTitle(String title) {
        this.mAppTitle = title;
    }

    public int getLaunchColor() {
        return this.mLaunchColor;
    }

    public boolean isTranslucentTheme() {
        return this.mIsSettingsItem;
    }

    public int getPriority() {
        return this.mPriority;
    }

    public void setPriority(int priority) {
        this.mPriority = priority;
    }

    public int getSettingsType() {
        return this.mSettingsType;
    }

    public void setSettingsType(int mSettingsType) {
        this.mSettingsType = mSettingsType;
    }

    public String getPackageName() {
        return this.mPackageName;
    }

    public String getInstallProgressString(Context context) {
        if (this.mInstallProgressPercent == -1) {
            return "";
        }
        return context.getString(R.string.progress_percent, this.mInstallProgressPercent);
    }

    public int getInstallProgressPercent() {
        return this.mInstallProgressPercent;
    }

    public String getInstallStateString(Context context) {
        return context.getString(this.mInstallStateStringResourceId);
    }

    public int getInstallStateStringResId() {
        return this.mInstallStateStringResourceId;
    }

    public void setInstallProgressPercent(int progressPercent) {
        this.mInstallProgressPercent = progressPercent;
    }

    public void setInstallStateStringResourceId(int stateStringResourceId) {
        this.mInstallStateStringResourceId = stateStringResourceId;
    }

    public boolean isInstalling() {
        return this.mInstallStateStringResourceId != 0;
    }

    public boolean isInitialInstall() {
        return this.mIsInitialInstall;
    }

    public long getFirstInstallTime() {
        return this.mPackageInstallTime;
    }

    public String toString() {
        return this.mAppTitle + " [" + this.mPackageName + "]";
    }

    public void cancelPendingOperations(Context context) {
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof LaunchPoint)) return false;

        LaunchPoint that = (LaunchPoint) o;

        if (mComponentName != null && !mComponentName.equals(that.mComponentName)) {
            return false;
        }

        return (mPackageName != null && !mPackageName.equals(that.mPackageName));
    }

    @Override
    public int hashCode() {
        int result = mComponentName != null ? mComponentName.hashCode() : 0;
        result = 31 * result + (mPackageName != null ? mPackageName.hashCode() : 0);
        return result;
    }
}

File: app/src/main/java/com/amazon/tv/leanbacklauncher/apps/LaunchPoint.kt
package com.amazon.tv.leanbacklauncher.apps

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.content.pm.ResolveInfo
import android.graphics.Bitmap
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import android.text.TextUtils
import androidx.core.content.ContextCompat
import androidx.core.content.res.ResourcesCompat
import androidx.palette.graphics.Palette
import com.amazon.tv.firetv.leanbacklauncher.apps.AppCategory
import com.amazon.tv.firetv.leanbacklauncher.util.AppCategorizer
import com.amazon.tv.firetv.leanbacklauncher.util.BannerUtil
import com.amazon.tv.firetv.leanbacklauncher.util.SettingsUtil.SettingsType
import com.amazon.tv.leanbacklauncher.R
import com.amazon.tv.leanbacklauncher.util.Util
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.request.RequestOptions
import com.bumptech.glide.request.target.CustomTarget
import com.bumptech.glide.request.transition.Transition
import java.util.*

class LaunchPoint {
    var title: String? = null
    var bannerDrawable: Drawable? = null
        private set
    var componentName: String? = null
        private set
    var contentDescription: String? = null
    private var mHasBanner = false
    var iconDrawable: Drawable? = null
    var installProgressPercent = 0
    var installStateStringResId = 0
        private set
    private var isInitialInstall = false
    var launchColor = 0
        private set
    var launchIntent: Intent? = null
        private set
    private var mListener: InstallingLaunchPointListener? = null
    var firstInstallTime: Long = 0
        private set
    var packageName: String? = null
        private set
    var priority = 0
    var settingsType = SettingsType.UNKNOWN.code
    private var mTranslucentTheme = false
    var isTranslucentTheme = false
        private set
    var appCategory: AppCategory? = null
        private set

    constructor()
    internal constructor(context: Context, appTitle: String?, packageName: String?) {
        clear(context)
        title = appTitle
        this.packageName = packageName
    }

    constructor(
        context: Context,
        appTitle: String?,
        iconUrl: String?,
        pkgName: String?,
        launchIntent: Intent?,
        listener: InstallingLaunchPointListener?
    ) {
        title = appTitle
        val resources = context.resources
        launchColor =
            ResourcesCompat.getColor(resources, R.color.app_launch_ripple_default_color, null)
        packageName = pkgName
        firstInstallTime = Util.getInstallTimeForPackage(context, packageName)
        if (launchIntent != null) {
            this.launchIntent =
                launchIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_RESET_TASK_IF_NEEDED)
            if (this.launchIntent!!.component != null) {
                componentName = this.launchIntent!!.component!!.flattenToString()
            }
        }
        mListener = listener
        isInitialInstall = true
    }

    constructor(
        context: Context,
        appTitle: String?,
        iconUrl: String?,
        launchIntent: Intent?,
        launchColor: Int,
        listener: InstallingLaunchPointListener?
    ) {
        clear(context)
        title = appTitle
        this.launchColor = launchColor
        if (launchIntent != null) {
            this.launchIntent =
                launchIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_RESET_TASK_IF_NEEDED)
            if (this.launchIntent!!.component != null) {
                componentName = this.launchIntent!!.component!!.flattenToString()
                packageName = this.launchIntent!!.component!!.packageName
            }
        }
        val resources = context.resources
        this.launchColor =
            ResourcesCompat.getColor(resources, R.color.app_launch_ripple_default_color, null)
        firstInstallTime = Util.getInstallTimeForPackage(context, packageName)
        mListener = listener
        isInitialInstall = true
        if (!TextUtils.isEmpty(iconUrl)) {
            val maxIconSize = resources.getDimensionPixelSize(R.dimen.banner_icon_size)
            Glide.with(context).asDrawable()
                .apply(RequestOptions.skipMemoryCacheOf(true))
                .apply(RequestOptions.diskCacheStrategyOf(DiskCacheStrategy.NONE))
                .load(iconUrl)
                .into(object : CustomTarget<Drawable?>(maxIconSize, maxIconSize) {
                    override fun onResourceReady(
                        resource: Drawable,
                        transition: Transition<in Drawable?>?
                    ) {
                        val launchPointListener = mListener
                        launchPointListener?.let {
                            iconDrawable = resource
                            it.onInstallingLaunchPointChanged(this@LaunchPoint)
                        }
                    }

                    override fun onLoadCleared(placeholder: Drawable?) {
                    }
                })
        }
    }

    // TODO
    constructor(
        context: Context,
        appTitle: String?,
        iconDrawable: Drawable?,
        launchIntent: Intent?,
        launchColor: Int
    ) {
        clear(context)
        title = appTitle
        this.iconDrawable = iconDrawable
        this.launchColor = launchColor
        if (launchIntent != null) {
            this.launchIntent =
                launchIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_RESET_TASK_IF_NEEDED)
            if (this.launchIntent!!.component != null) {
                componentName = this.launchIntent!!.component!!.flattenToString()
                packageName = this.launchIntent!!.component!!.packageName
                firstInstallTime = Util.getInstallTimeForPackage(context, packageName)
            }
        }
    }

    constructor(ctx: Context, pm: PackageManager?, info: ResolveInfo) {
        set(ctx, pm, info)
        isTranslucentTheme = true
    }

    constructor(
        ctx: Context,
        pm: PackageManager?,
        info: ResolveInfo,
        useBanner: Boolean,
        settingsType: Int
    ) {
        set(ctx, pm, info, useBanner)
        this.settingsType = settingsType
        isTranslucentTheme = true
    }

    @JvmOverloads
    fun set(
        ctx: Context,
        pm: PackageManager?,
        info: ResolveInfo,
        useBanner: Boolean = true
    ): LaunchPoint {
        clear(ctx)
        title = info.loadLabel(pm).toString()
        launchIntent = getLaunchIntent(info)
        launchIntent?.component?.let {
            componentName = it.flattenToString()
            packageName = it.packageName
        }
        val res = ctx.resources
        val maxWidth = res.getDimensionPixelOffset(R.dimen.max_banner_image_width)
        val maxHeight = res.getDimensionPixelOffset(R.dimen.max_banner_image_height)
        val activityInfo = info.activityInfo
        activityInfo?.let { ai ->
            appCategory = AppCategorizer.getAppCategory(packageName, ai)
            if (useBanner) {
                bannerDrawable = ai.loadBanner(pm)
                if (bannerDrawable is BitmapDrawable) {
                    bannerDrawable = BitmapDrawable(
                        res,
                        Util.getSizeCappedBitmap(
                            (bannerDrawable as BitmapDrawable?)!!.bitmap,
                            maxWidth,
                            maxHeight
                        )
                    )
                }
            }
            val overrides = BannerUtil.BANNER_OVERRIDES
            for (str in overrides.keys) {
                if (packageName!!.lowercase(Locale.getDefault()).contains(str)) {
                    bannerDrawable = ContextCompat.getDrawable(ctx, overrides[str]!!)
                    if (bannerDrawable is BitmapDrawable) {
                        bannerDrawable = BitmapDrawable(
                            res,
                            Util.getSizeCappedBitmap(
                                (bannerDrawable as BitmapDrawable?)!!.bitmap,
                                maxWidth,
                                maxHeight
                            )
                        )
                    }
                    break
                }
            }
            if (bannerDrawable != null) {
                mHasBanner = true
            } else {
                if (useBanner) {
                    bannerDrawable = ai.loadLogo(pm)
                    if (bannerDrawable is BitmapDrawable) {
                        bannerDrawable = BitmapDrawable(
                            res,
                            Util.getSizeCappedBitmap(
                                (bannerDrawable as BitmapDrawable?)!!.bitmap,
                                maxWidth,
                                maxHeight
                            )
                        )
                    }
                }
                if (bannerDrawable != null) {
                    mHasBanner = true
                } else {
                    mHasBanner = false
                    iconDrawable = info.loadIcon(pm)
                }
            }
        }
        priority = info.priority
        mTranslucentTheme = isTranslucentTheme(ctx, info)
        launchColor = getColor(ctx, info)
        firstInstallTime = Util.getInstallTimeForPackage(ctx, packageName)
        return this
    }

    fun addLaunchIntentFlags(flags: Int) {
        launchIntent?.addFlags(flags)
    }

    private fun clear(context: Context) {
        cancelPendingOperations(context)
        installProgressPercent = -1
        installStateStringResId = 0
        componentName = null
        packageName = null
        bannerDrawable = null
        title = null
        contentDescription = null
        iconDrawable = null
        launchColor = 0
        launchIntent = null
        mHasBanner = false
        priority = -1
        settingsType = SettingsType.UNKNOWN.code
        isInitialInstall = false
        mListener = null
        firstInstallTime = -1
    }

    fun setInstallationState(launchPoint: LaunchPoint): LaunchPoint {
        installProgressPercent = launchPoint.installProgressPercent
        installStateStringResId = launchPoint.installStateStringResId
        return this
    }

    fun hasBanner(): Boolean {
        return mHasBanner
    }

    val isGame: Boolean
        get() = appCategory == AppCategory.GAME

    fun getInstallProgressString(context: Context): String {
        return if (installProgressPercent == -1) {
            ""
        } else context.getString(R.string.progress_percent, installProgressPercent)
    }

    fun getInstallStateString(context: Context): String {
        return context.getString(installStateStringResId)
    }

    fun setInstallStateStringResourceId(stateStringResourceId: Int) {
        installStateStringResId = stateStringResourceId
    }

    val isInstalling: Boolean
        get() = installStateStringResId != 0

    override fun toString(): String {
        return "$title [$packageName]"
    }

    fun dump(): String {
        return "$title [$packageName]\n" +
                "component:$componentName\n" +
                "contentDescription:$contentDescription\n" +
                "hasBanner:$mHasBanner\n" +
                "launchColor:$launchColor\n" +
                "launchIntent:$launchIntent\n" +
                "priority:$priority\n" +
                "settingsType:$settingsType\n" +
                "isTranslucentTheme:$isTranslucentTheme\n" +
                "isGame:$isGame\n" +
                "appCategory:$appCategory\n"
    }

    fun cancelPendingOperations(context: Context?) {}

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is LaunchPoint) return false
        return if (componentName != null && componentName != other.componentName) {
            false
        } else packageName != null && packageName != other.packageName
    }

    override fun hashCode(): Int {
        var result = if (componentName != null) componentName.hashCode() else 0
        result = 31 * result + if (packageName != null) packageName.hashCode() else 0
        return result
    }

    companion object {
        private fun getLaunchIntent(info: ResolveInfo): Intent {
            val componentName =
                ComponentName(info.activityInfo.applicationInfo.packageName, info.activityInfo.name)
            val intent = Intent("android.intent.action.MAIN")
            intent.addCategory("android.intent.category.LAUNCHER")
            intent.component = componentName
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_RESET_TASK_IF_NEEDED)
            return intent
        }

        private fun getColor(myContext: Context, info: ResolveInfo): Int {
            return try {
                val theme = myContext.createPackageContext(
                    info.activityInfo.applicationInfo.packageName,
                    0
                ).theme
                theme.applyStyle(info.activityInfo.themeResource, true)
                val a = theme.obtainStyledAttributes(intArrayOf(android.R.attr.colorPrimary))
                val color = a.getColor(
                    0,
                    ResourcesCompat.getColor(myContext.resources, R.color.banner_background, null)
                )
                a.recycle()
                color
            } catch (e: PackageManager.NameNotFoundException) {
                e.printStackTrace()
                ResourcesCompat.getColor(myContext.resources, R.color.banner_background, null)
            }
        }

        private fun getColor(context: Context, bitmap: Bitmap): Int {
            val p = Palette.from(bitmap).generate()
            var swatch = p.darkMutedSwatch
            if (swatch != null) return swatch.rgb else if (p.lightMutedSwatch.also {
                    swatch = it
                } != null) return swatch!!.rgb else if (p.vibrantSwatch.also {
                    swatch = it
                } != null) return swatch!!.rgb else if (p.mutedSwatch.also {
                    swatch = it
                } != null) return swatch!!.rgb else if (p.darkVibrantSwatch.also {
                    swatch = it
                } != null) return swatch!!.rgb else if (p.darkMutedSwatch.also {
                    swatch = it
                } != null) return swatch!!.rgb
            return ContextCompat.getColor(context, R.color.banner_background) // for API 22
        }

        fun isTranslucentTheme(myContext: Context, info: ResolveInfo): Boolean {
            return try {
                val theme = myContext.createPackageContext(
                    info.activityInfo.applicationInfo.packageName,
                    0
                ).theme
                theme.applyStyle(info.activityInfo.themeResource, true)
                val a = theme.obtainStyledAttributes(intArrayOf(android.R.attr.windowIsTranslucent))
                val windowIsTranslucent = a.getBoolean(0, false)
                a.recycle()
                windowIsTranslucent
            } catch (e: PackageManager.NameNotFoundException) {
                e.printStackTrace()
                false
            }
        }
    }
}